### DISCUSSION TITLE
Needed clarification on how number of queues are defined in conductor

### DISCUSSION THREAD

**Discussion by send2mridul** [Netflix/conductor#2815] [Q&A] [answered]
Hello everyone, I am a beginner to **Conductor workflow** and needed clarification on below.

If I have 3 instances of same workflow and all are ran together as follows.

Workflow Instance 1 - 
<img width="275" alt="image" src="https://user-images.githubusercontent.com/75941027/156802418-c42af850-5b5c-4eb1-a65f-2e61f9a671b2.png">

Workflow Instance 2 - 
<img width="275" alt="image" src="https://user-images.githubusercontent.com/75941027/156802422-ae569b4b-7853-4fd6-9b98-aaa3d893669b.png">

Workflow Instance 3 - 
<img width="275" alt="image" src="https://user-images.githubusercontent.com/75941027/156802424-d1fc58a8-2f72-4bef-8eeb-557f7bb0a405.png">

Since its a worker task, all tasks will be pushed into the queue : [W3-T1, W2-T1, W1-T1]
**(W - Workflow instance, T - Task no)**

Now, when I poll once for the first task i.e "verify_if_idents_are_added_test", I get the task of 1st workflow instance. I do post call to update the result of same.

Now, following is the current state of workflow instances : 

Workflow Instance 1 -
<img width="279" alt="image" src="https://user-images.githubusercontent.com/75941027/156803347-895d0d04-1b19-4b77-8bcc-b48f1de4f032.png">

Workflow Instance 2 - 
<img width="275" alt="image" src="https://user-images.githubusercontent.com/75941027/156802422-ae569b4b-7853-4fd6-9b98-aaa3d893669b.png">

Workflow Instance 3 - 
<img width="275" alt="image" src="https://user-images.githubusercontent.com/75941027/156802424-d1fc58a8-2f72-4bef-8eeb-557f7bb0a405.png">

Here, the 2nd task of 1st workflow instance is also pushed to the queue.
Our queue looks something like this : 
[W1-T2, W3-T1, W2-T1]

At this point, If I try to poll for Task2, I should be not getting W1-T2 as its not the first element of our queue. But I tried this and I got this task i.e 2nd task of first workflow instance. 

This means we don't have single queue or the organisation of queue is something different which I am not able to understand.

Can anyone help me understand how the no of queues are defined in conductor workflows. 

Thanks in advance !!

**Comment by v1r3n** [comment, marked as answer]
Hi 

the queues are maintained per workflow to guarantee the ordering of execution _within_ the workflow.  The behavior you see is as expected.  But if you have a specific use case where you expect something different, would love to understand your use case and see how to best address that in Conductor.

**Reply by send2mridul** [reply]
Hey @v1r3n, by workflow do you mean workflow instance ? I have different workflow instances of same workflow, so each will be having their own queue or there will be only one queue for one workflow which will be shared by all the instance of it ?

**Reply by rishabhjha708** [reply]
Hi @v1r3n , if its per workflow then in above example only 1 queue was there right ? then how workflow instance 1 moved to task 2 was scheduled and when polled via worker got that task, when already for workflow instance 2 and 3, "verify_if_ident_is_added_test" was already there in conductor queue.

**Reply by v1r3n** [reply]
Yes, I mean workflow instance. The queues are maintained per task type, so there are separate queues per task and they are not strictly FIFO - as these are time delay queues (ie you can put a message in a queue which may not come out until a certain time period).    

There are also queues per workflow that are used internally by Conductor for various housekeeping process.

So if you have a workflow with 3 tasks - T1, T2 and T3, there are 3 queues (one for each of the task).  For each workflow instance, the messages are pushed to these queues as the tasks get scheduled (so for example when a Workflow instance  completes the task 1  the task 2 is scheduled inside the queue for task 2 and so forth).

Hope this helps.

**Reply by v1r3n** [reply]
@rishabhjha708 see above - the queues are per task not per workflow.  Let us know if helps.

**Reply by send2mridul** [reply]
Got it! Thank you very much @v1r3n

**Reply by rishabhjha708** [reply]
Yes, @v1r3n  this helps. Thanks for your detailed explanation.
